File: Clone Graph - Leetcode 133/Clone Graph - Leetcode 133.py
from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if not node:
            return None

        start = node
        o_to_n = {}
        stk = [start]
        visited = set()
        visited.add(start)

        while stk:
            node = stk.pop()
            o_to_n[node] = Node(val=node.val)

            for nei in node.neighbors:
                if nei not in visited:
                    visited.add(nei)
                    stk.append(nei)

        for old_node, new_node in o_to_n.items():
            for nei in old_node.neighbors:
                new_nei = o_to_n[nei]
                new_node.neighbors.append(new_nei)

        return o_to_n[start]

# Time Complexity: O(V+E)
# Space Complexity: O(V)

File: Clone Graph - Leetcode 133/Clone Graph - Leetcode 133.java
import java.util.*;

public class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;

        Map<Node, Node> oToN = new HashMap<>();
        Stack<Node> stk = new Stack<>();
        Set<Node> visited = new HashSet<>();

        stk.push(node);
        visited.add(node);

        // DFS to create the nodes
        while (!stk.isEmpty()) {
            Node curr = stk.pop();
            oToN.put(curr, new Node(curr.val));

            for (Node nei : curr.neighbors) {
                if (!visited.contains(nei)) {
                    visited.add(nei);
                    stk.push(nei);
                }
            }
        }

        // DFS to set up neighbors
        for (Map.Entry<Node, Node> entry : oToN.entrySet()) {
            Node oldNode = entry.getKey();
            Node newNode = entry.getValue();
            for (Node nei : oldNode.neighbors) {
                newNode.neighbors.add(oToN.get(nei));
            }
        }

        return oToN.get(node);
    }
}


